### PR TITLE
Add --burst-limit option for client-side throttling limit configuration

### LINKED ISSUE #10560: Throttling messages while running helm upgrade commands
## Description of problem:

User facing throttling messages when running the `./helm upgrade` command. The same issue with `oc` commands as well but after the `oc` binary upgrade issue is solved. Need a similar fix in helm binary as well.

Find this bug for `oc` client: https://bugzilla.redhat.com/show_bug.cgi?id=1899575

## How reproducible:
- Intermittent

## Steps to Reproduce:
1. Add many CRD's (>100) to OpenShift 4 and add workloads
2. Run helm upgrade commands on the cluster

## Actual results:
- Throttling messages while running.
```
[root@Server~]# ./helm upgrade --debug --namespace helm-example --install --set env.active_profile=dev,environment=dev --wait platform-auth-internal-api https://example.com/artifactory/helm-local/helm-example/platform-auth-internal-api-394419.tgz --dry-run -v 6  > /tmp/test
history.go:56: [debug] getting history for release platform-auth-internal-api
install.go:173: [debug] Original chart version: ""
install.go:190: [debug] CHART PATH: /root/.cache/helm/repository/platform-auth-internal-api-394419.tgz

I1028 11:37:01.745973   94954 request.go:668] Waited for 1.168719201s due to client-side throttling, not priority and fairness, request: GET:https://api.example.com:6443/apis/apps.openshift.io/v1?timeout=32s
```

## Expected results:
- No throttling messages while running it. 

## Output of `helm version`:
`version.BuildInfo{Version:"v3.6.2+5.el8", GitCommit:"eb607dd4f123eaedab662cef21008d177f2c3426", GitTreeState:"clean", GoVersion:"go1.15.13"}`

## Output of `kubectl version`:
`N/A`

## Cloud Provider/Platform (AKS, GKE, Minikube etc.): 
`OpenShift`

### PR DESCRIPTION
**Pull request by isutton**
**What this PR does / why we need it**:

Client-side throttling seems to be an issue in larger environments such as OpenShift clusters, where it is common to have several hundreds CRDs out-of-the-box.

From this view point, it is fair that clients should be able to fine tune this accordingly should the environment they work on evolves, which is currently not possible, and quite frustrating.

This change introduces the --burst-limit option to helm (and its counterpart HELM_BURST_LIMIT environment variable) to address that issue, allowing clients to properly tune their client usage as their environment evolves.

**Special notes for your reviewer**:

**If applicable**:
- [x] this PR contains documentation
- [x] this PR contains unit tests
- [ ] this PR has been tested for backwards compatibility

Closes #10560

### CODE REVIEW

**Comment by yxxhero** [comment]
@isutton Please create a issue this PR. Thanks very much.

**Comment by yxxhero** [comment]
--default-burst-limit   --> --burst-limit or --kube-burst-limit
when --burst-limit not set. it will be set defaultBurstLimit.

**Comment by isutton** [comment]
> ` [75398fc](/helm/helm/pull/10842/commits/75398fca3dc7d7adefd242fbf6815994faaa1778)`

Fixed in 75398fc.

**Comment by isutton** [comment]
> @isutton Please create a issue this PR. Thanks very much.

This PR closes #10560.

File: pkg/cli/environment.go [hunk 1]
```diff
@@ -30,13 +30,17 @@ import (
...
+	"k8s.io/client-go/rest"
 
 	"helm.sh/helm/v3/pkg/helmpath"
 )
 
 // defaultMaxHistory sets the maximum number of releases to 0: unlimited
 const defaultMaxHistory = 10
 
+// defaultBurstLimit sets the default client-side throttling limit
+const defaultBurstLimit = 250
```
Comment (yxxhero) [inline pkg/cli/environment.go:42]: why is 250?

**Comment by yxxhero** [comment]
unittest?

File: pkg/cli/environment.go (hunk 1 above)
Comment (isutton) [reply pkg/cli/environment.go:42]: This is the same value OpenShift's `oc` is using as default, see https://github.com/helm/helm/issues/10560.

**Comment by isutton** [comment]
> unittest?

There are no unit tests as the value is only kept to be forwarded to the client using the appropriate API; additionally there are evidences through the change in `oc` that this value works reasonably well with OpenShift out-of-the-box.

File: pkg/cli/environment.go (hunk 1 above)
Comment (yxxhero) [reply pkg/cli/environment.go:42]: ok

File: cmd/helm/root.go [hunk 2]
```diff
@@ -67,6 +67,7 @@ Environment variables:
 | $HELM_KUBEASUSER                   | set the Username to impersonate for the operation.                                |
 | $HELM_KUBECONTEXT                  | set the name of the kubeconfig context.                                           |
 | $HELM_KUBETOKEN                    | set the Bearer KubeToken used for authentication.                                 |
+| $HELM_DEFAULT_BURST_LIMIT          | set the default burst limit in the case the server contains many CRDs             |
```
Comment (yxxhero) [inline cmd/helm/root.go:70]: HELM_DEFAULT_BURST_LIMIT  --> HELM_BURST_LIMIT is right?

**Comment by yxxhero** [comment]
add your test case into TestEnvSettings @isutton

**Comment by yxxhero** [comment]
pkg/cli/environment_test.go

File: cmd/helm/root.go (hunk 2 above)
Comment (isutton) [reply cmd/helm/root.go]: This has been fixed.

**Comment by isutton** [comment]
> add your test case into TestEnvSettings @isutton

TestEnvSettings has been amended to consider HELM_BURST_LIMIT environment variable and --burst-limit option.

**Comment by yxxhero** [comment]
@isutton I think all is ok. need a core maintainer to review this.

**Comment by yxxhero** [comment]
@isutton I think all is ok. need a core maintainer to review this.

File: cmd/helm/root.go [hunk 3]
```diff
@@ -67,6 +67,7 @@ Environment variables:
 | $HELM_KUBEASUSER                   | set the Username to impersonate for the operation.                                |
 | $HELM_KUBECONTEXT                  | set the name of the kubeconfig context.                                           |
 | $HELM_KUBETOKEN                    | set the Bearer KubeToken used for authentication.                                 |
+| $HELM_BURST_LIMIT                  | set the default burst limit in the case the server contains many CRDs             |
```
Comment (sabre1041) [inline cmd/helm/root.go:70]: Let's both call out the default value as well how it can be disabled (-1)

File: pkg/cli/environment.go (hunk 1 above)
Comment (sabre1041) [inline pkg/cli/environment.go:42]: Recommend setting the value to `100` as the default. The default in the Kubernetes rest client is `10` and `250` seems to be a dramatic change

**Review by sabre1041 — CHANGES_REQUESTED** [review]
This looks good. A few comments

File: cmd/helm/root.go (hunk 3 above)
Comment (isutton) [reply cmd/helm/root.go]: Fixed in https://github.com/helm/helm/pull/10842/commits/14daf80ca0301d987636e66a4b6d70452389b92c

File: pkg/cli/environment.go (hunk 1 above)
Comment (isutton) [reply pkg/cli/environment.go]: Fixed in https://github.com/helm/helm/pull/10842/commits/37608533d4f700b7281eff3e93f8dcf87ec5989b

**Review by sabre1041 — APPROVED** [review]
LGTM

**Pull request merged**